### ISSUE TITLE
Design thoughts: header/trailer advanced usage

### ISSUE THREAD

**Issue by mgravell** [protobuf-net/protobuf-net.Grpc#285]
So ties into #284

Consider:

``` c#
SomeType Method1(SomeType arg, [Header] string x, ...)
SomeType Method2(SomeType arg, [Header] out string x, ...)
SomeType Method3(SomeType arg, [Header] ref string x, ...)
SomeType Method1(SomeType arg, [Trailer] string x, ...)
SomeType Method2(SomeType arg, [Trailer] out string x, ...)
SomeType Method3(SomeType arg, [Trailer] ref string x, ...)
```

- [Header] also interchangeable with aspnet [FromHeader]
- optional name, otherwise uses arg name
- types: limited; ala Convert? always invariant; suggest string, unsigned+signed integers 8/16/32/64, byte[], bool, char, DateTime
- Directionality:
  - nothing/in: client-to server
  - out: server-to-client
  - ref: both directions
- [Trailer] not allowed for C2S/both for client-streaming/duplex
- [Trailer] not allowed for S2C/both for server-streaming/duplex

Thoughts?

**Comment by menaheme** [comment]
Does this open up the way to have more than one input parameter (excluding context)?

**Comment by mgravell** [comment]
Yes, although I also want to allow full body binding as protobuf implied models. But I have definitely missed an important category above - binary headers that are serialized object payloads. I'm conflicted as to whether to add this one, as header sizes are bounded. Open to thoughts, though... i.e.

```
..., [Header] SomeType arg, ...
..., [Header] ref SomeType arg, ...
..., [Header] out SomeType arg, ...
```

(and [Trailer])

which should only work if `SomeType` is recognized by the configured marshaller

**Comment by menaheme** [comment]
Hmm, the difference between method parameter types and header types, is that the former are contracted in the interface signature , the latter are not documented anywhere, which is risky

**Comment by mgravell** [comment]
Partly. Also no streaming for headers. And headers have more limited sizes. And the marshaller API is different (although I've suggested some improvements for that over on gRPC.Core)

**Comment by menaheme** [comment]
How does one communicate to comsumer, the header type expected at the server?

**Comment by mgravell** [comment]
Well, in code-first land: like the above :)

In contract-first: "you tell them"

**Comment by sethsnel** [comment]
TL;DR; Would this design proposal fix our usecase?

We have an usecase where we want to read some property from general metadata. The server expects the client to ship this metadata key/value across requests. (we use an intercepter in the client to inject this metadata).

On the server we have a method to extract metadata:
![image](https://github.com/user-attachments/assets/3ddfb2fb-3017-4bc2-90a7-e2f8a50412a8)

It needs servercellcontext in the specific method:
![image](https://github.com/user-attachments/assets/e2c3abb8-a8a7-4d92-88bc-73929e83c5c8)

Now it obfuscates the interface definition:
![image](https://github.com/user-attachments/assets/0e192b51-2e3b-4de2-b853-f071bbe55fcd)

Would this design proposal fix our usecase? @mgravell

**Comment by mgravell** [comment]
Note: you should usually prefer CallContext rather than ServerCallContext so that the same API can be used by client and server. Is it the case, then, that adding CallContext is unwanted in your case?

**Comment by sethsnel** [comment]
In this approach it becomes part of the interface, that's our struggle. Since we use the code-first approach, it means that the client needs to provide the `CallContext` on every method call. I would like to find out someway in which the client can insert metadata that is sent along with every method call. And on the server side that can be read (without changes the shared interface definition). Do you perhaps know a direction or some documentation on this?

**Comment by mgravell** [comment]
You could almost certainly do this with an interceptor, with the current libraries - and it would work with vanilla gRPC (both Google and Microsoft transports) and protobuf-net.Grpc; want an example?

**Comment by sethsnel** [comment]
Funny you should mention that. We're already using interceptors, but I have two questions about them.
- How to set metadata on the client side? (We now read from httpContextAccessor but I'd rather prefer if the metadata could be set within a function call context)
- How to read metadata within server implementation method? (we now inject a scoped 'ContextRepository' inside the incepter, extract the metadata. Then in the server implemantion method, we read the values we need from the same contextrepostory instance. This seems implementation feels cumbersome)

Current ClientInterceptor to set metadata:
![image](https://github.com/user-attachments/assets/7d20f14b-35c5-478a-897d-06fff18a9606)

Current Interceptor to read on server side:
![image](https://github.com/user-attachments/assets/c3e4b402-6d6a-4425-8da6-973ad71e60fc)
![image](https://github.com/user-attachments/assets/647650d8-7226-47fd-bda2-8954dfe63c00)

**Comment by mgravell** [comment]
This is partly why using the metadata explicitly is the simplest approach here, even if it means adding a CallContext - that works at both ends fairly clearly. Without that: maybe AsyncLocal? I know that should work client-side; server-side, I'd need to check the marshalling layers to check we rent on the wrong side of a dispatch barrier.

**Comment by mgravell** [comment]
I also wonder if @JamesNK might have thoughts here... especially if we just focus on the core gRPC bits; ambient state, transfer to server, etc. Any recommended practices here?

**Comment by JamesNK** [comment]
Thoughts:
1. Headers can have multiple values. If there are multiple you could either throw an error or concat values together with a comma. Maybe if someone specifies `[Header] string[] x` then multi values would be added to the array. ASP.NET Core uses `StringValues` type for this.
2. It's common to have hyphens in headers. Would need a way to specify a header name in the attribute, e.g. `[Header("x-my-header")] string x` (edit: I just noticed you mentioned this feature)
3. Binary headers are a thing (`-bin` suffix). Would need to think about how to handle them. Require that the type be byte array?
4. It would be weird that you could access incoming headers, and output outgoing trailers, but not outgoing headers. The reason is because outgoing headers are sent with the first message in a stream, but people using the library wouldn't know that. Maybe just limit feature to incoming headers.

ASP.NET Core has `[FromHeader]`. It's basically the same feature. Look at what it does to help design this.
https://dotnettutorials.net/lesson/fromheader-in-asp-net-core-web-api/

**Comment by mgravell** [comment]
@JamesNK so to be clear: in that scenario we're back in the "protobuf-net.Grpc adds interpretation and wrangling" scenario - there's no great inbuilt / tested built in mechanisms for conveying additional state via headers?

**Comment by mgravell** [comment]
My other thought here: I need to switch to using a generator approach; runtime inspection and emit is too brittle as we add increasing features and other complexity. It would be better to accelerate that work and add these features in generated code, than to keep hacking away at the increasingly overweight runtime code.
